File: leet_lint_code/lint_code/binary_tree_level_order_traversal.cpp
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of lint code problem 69.

#ifndef LEETLINTCODE_LINTCODE_BINARYTREELEVELORDERTRAVERSAL_CPP
#define LEETLINTCODE_LINTCODE_BINARYTREELEVELORDERTRAVERSAL_CPP

#include <vector>

class TreeNode {
 public:
   TreeNode(int val) {
     this->val = val;
     this->left = this->right = NULL;
   }

   int val;
   TreeNode *left, *right;
};

class BinaryTreeLevelOrderTraversal {
 public:
   vector<vector<int>> levelOrder(TreeNode* root) {

   }
};

#endif // LEETLINTCODE_LINTCODE_BINARYTREELEVELORDERTRAVERSAL_CPP

File: leet_lint_code/leet_code/java/BinaryTreeLevelOrderTraversal.java
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of leet code problem 102.
// Date: 2019.8.14

import java.util.*;

public class BinaryTreeLevelOrderTraversal {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        Queue<TreeNode> queue1 = new LinkedList<>();
        Queue<TreeNode> queue2 = new LinkedList<>();
        if (root == null) {
            return ans;
        }

        queue1.add(root);
        while (!queue1.isEmpty() || !queue2.isEmpty()) {
             List<Integer> levelAns = new ArrayList<>();

             if (queue1.isEmpty()) {
                 while (!queue2.isEmpty()) {
                     TreeNode head = queue2.poll();
                     if (head.left != null) {
                         queue1.add(head.left);
                     }

                     if (head.right != null) {
                         queue1.add(head.right);
                     }
                     levelAns.add(head.val);
                 }
             } else {
                 while (!queue1.isEmpty()) {
                     TreeNode head = queue1.poll();
                     if (head.left != null) {
                         queue2.add(head.left);
                     }

                     if (head.right != null) {
                         queue2.add(head.right);
                     }
                     levelAns.add(head.val);
                 }
             }
             ans.add(levelAns);
        }

        return ans;
    }
}
